compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -307,46 +307,47 @@ fn negative_impl<'cx, 'tcx>(
-                &infcx,
-                FulfillmentContext::new(),
-                ObligationCause::dummy(),
-                impl1_env,
-                impl1_trait_ref,
-            ) {
-                Ok(impl1_trait_ref) => impl1_trait_ref,
-                Err(err) => {
-                    bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
-                }
-            };
-
-            // Attempt to prove that impl2 applies, given all of the above.
-            let selcx = &mut SelectionContext::new(&infcx);
-            let impl2_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl2_def_id);
-            let (impl2_trait_ref, obligations) =
-                impl_trait_ref_and_oblig(selcx, impl1_env, impl2_def_id, impl2_substs);
-
-            !obligations_satisfiable(
-                &infcx,
-                impl1_env,
-                impl1_def_id,
-                impl1_trait_ref,
-                impl2_trait_ref,
-                obligations,
-            )
-        } else {
-            let ty1 = tcx.type_of(impl1_def_id);
-            let ty2 = tcx.type_of(impl2_def_id);
-
-            !obligations_satisfiable(&infcx, impl1_env, impl1_def_id, ty1, ty2, iter::empty())
+        match tcx.impl_header(impl1_def_id) {
+            ImplSubject::Trait(impl1_trait_ref) => {
+                // Normalize the trait reference. The WF rules ought to ensure
+                // that this always succeeds.
+                let impl1_trait_ref = match traits::fully_normalize(
+                    &infcx,
+                    FulfillmentContext::new(),
+                    ObligationCause::dummy(),
+                    impl1_env,
+                    impl1_trait_ref,
+                ) {
+                    Ok(impl1_trait_ref) => impl1_trait_ref,
+                    Err(err) => {
+                        bug!("failed to fully normalize {:?}: {:?}", impl1_trait_ref, err);
+                    }
+                };
+
+                // Attempt to prove that impl2 applies, given all of the above.
+                let selcx = &mut SelectionContext::new(&infcx);
+                let impl2_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl2_def_id);
+                let (impl2_trait_ref, obligations) =
+                    impl_trait_ref_and_oblig(selcx, impl1_env, impl2_def_id, impl2_substs);
+
+                !equate(
+                    &infcx,
+                    impl1_env,
+                    impl1_def_id,
+                    impl1_trait_ref,
+                    impl2_trait_ref,
+                    obligations,
+                )
+            }
+            ImplSubject::Inherent(ty1) => {
+                let ty2 = tcx.type_of(impl2_def_id);
+                !equate(&infcx, impl1_env, impl1_def_id, ty1, ty2, iter::empty())
+            }
         }
     })
 }
 
-fn obligations_satisfiable<'cx, 'tcx, T: Debug + ToTrace<'tcx>>(
+fn equate<'cx, 'tcx, T: Debug + ToTrace<'tcx>>(
```

Comment:
I was imagining this not taking a `T` but rather taking two `ImplSubject<'tcx>` arguments.